### ISSUE TITLE
tools that write BAM/CRAM should have option to index

### ISSUE THREAD

**Issue by keiranmraine** [samtools/samtools#774]
Tools that (can) write a sorted alignment file should have the ability to generate the index file while writing the output.

It would be preferable to have this than use `tee` to achieve this via pipes.

Reason: biobambam2 has this for most tools but the build time is quite considerable, samtools now has most of the functionality I need (for downstream processing) so I'm looking at reducing complexity of tools in docker images.

**Comment by daviesrob** [comment]
Agreed, it would be useful.  It would need changes to HTSlib though, as I don't think it supports indexing on-the-fly at the moment.

**Comment by keiranmraine** [comment]
Alternatively the ability to `--passthrough` the stream on tools (on view useful to do this even when writing to file), e.g.

```
samtools view -S -C -T ref.fa --passthrough -o new.cram some.sam | \
  samtools index --passthrough - new.cram.crai | \
    md5sum > new.cram.md5
```

(I know `-S` is ignored, just for extra clarity)

**Comment by daviesrob** [comment]
It should be possible to add a passthrough mode to `samtools index`, although we would effectively be implementing tee in it.  It would be easier than adding indexing to everything, which although useful would require ABI changes.

For this particular case you could try [teepot](https://github.com/wtsi-npg/teepot) which is mostly like tee, but better at handling cases where one of the output pipes either pauses for a long time or stops reading completely.

```
samtools view -S -C -T ref.fa some.sam | \
  teepot new.cram >(samtools index - new.cram.crai) >(md5sum)
```

**Comment by keiranmraine** [comment]
Ah, I see this works without any changes... thank you for the (re)education

**Comment by jkbonfield** [comment]
Depends on samtools/htslib#718, but will also need a small samtools tweak to use the new htslib code.

**Comment by nh13** [comment]
@jkbonfield bump

**Comment by jkbonfield** [comment]
Good point.  The htslib PR to do this was written a year ago but only arrived in Feb so I kind of forgot that there was samtools follow-up work waiting on it.  I'll try and remind myself what the grand plan was!

**Comment by jkbonfield** [comment]
It's a work in progress now so expect PR soon.  I decided to use the --long-opts global options style as we need indices specifying for every program that outputs.

In the first incarnation I'm going to support CSI and CRAI **only** with no parameterisation of CSI.  We can improve it later if needed, but given this is new functionality I don't see an issue with also requiring the improved index format too.  BAI should die really given the problems with it.  I guess if we must have it, we'll need another --index-format options, but that then entails all sorts of complications when writing mixed BAM and CRAM for example.  It just gets messy for both the user and me.

An example command line:

```
./samtools view --write-index 9827_2#49.100k.bam -U /tmp/a.u.bam -o /tmp/a.cram;ls -l /tmp/a.*
-rw-r--r-- 1 jkb team117 6073993 Jun  5 12:08 /tmp/a.cram
-rw-r--r-- 1 jkb team117     209 Jun  5 12:08 /tmp/a.cram.crai
-rw-r--r-- 1 jkb team117    6861 Jun  5 12:08 /tmp/a.u.bam
-rw-r--r-- 1 jkb team117      73 Jun  5 12:08 /tmp/a.u.bam.csi
```

So far `view` is the only sub-command I've done.

**Comment by valeriuo** [comment]
Closed by merging #1062 into develop
